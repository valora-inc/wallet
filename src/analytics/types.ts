export enum ScrollDirection {
  next = 'next',
  previous = 'previous',
}

export enum BackQuizProgress {
  word_chosen = 'word_chosen',
  backspace = 'backspace',
}

// Origin of Send actions
export enum SendOrigin {
  AppSendFlow = 'app_send_flow', // Sending as part of the app send flow
  AppRequestFlow = 'app_request_flow', // Sending because of a received payment request
  Bidali = 'bidali', // Sending from Bidali
}

// Origin of WalletConnect pairing
export enum WalletConnectPairingOrigin {
  Scan = 'scan',
  Deeplink = 'deeplink',
}

// Origin of WalletConnect/DappKit request
export enum DappRequestOrigin {
  InAppWebView = 'in_app_web_view',
  External = 'external',
}
