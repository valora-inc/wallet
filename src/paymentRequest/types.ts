import { NotificationTypes } from 'src/notifications/types'

export enum PaymentRequestStatus {
  REQUESTED = 'REQUESTED',
  COMPLETED = 'COMPLETED',
  DECLINED = 'DECLINED',
  CANCELLED = 'CANCELLED',
}

export interface PaymentRequest {
  uid?: string
  amount: string
  comment?: string
  createdAt: number
  requesterAddress: string
  requesteeAddress: string
  requesterE164Number?: string
  status: PaymentRequestStatus
  notified: boolean
  type?: NotificationTypes.PAYMENT_REQUESTED
}

export interface WriteablePaymentRequest extends Omit<PaymentRequest, 'createdAt'> {
  createdAt: object
}
