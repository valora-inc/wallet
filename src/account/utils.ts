import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
