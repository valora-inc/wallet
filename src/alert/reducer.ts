import { Action } from '@reduxjs/toolkit'
import { Actions, ActionTypes, AlertTypes } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { RootState } from 'src/redux/reducers'

export enum ErrorDisplayType {
  'BANNER',
  'INLINE',
}

export interface Alert {
  type: AlertTypes
  displayMethod: ErrorDisplayType
  message: string
  dismissAfter?: number | null
  buttonMessage?: string | null
  action?: Action | null
  title?: string | null
  underlyingError?: ErrorMessages | null
}

type State = Alert | null

const initialState = null

export const reducer = (state: State = initialState, action: ActionTypes): State => {
  switch (action.type) {
    case Actions.SHOW:
      return {
        displayMethod: action.displayMethod,
        type: action.alertType,
        message: action.message,
        dismissAfter: action.dismissAfter,
        buttonMessage: action.buttonMessage,
        action: action.action,
        title: action.title,
        underlyingError: action.underlyingError,
      }
    case Actions.HIDE:
      return null
    default:
      if (state?.action === action) {
        // Hide alert when the alert action is dispatched
        return null
      }
      return state
  }
}

export const errorSelector = (state: RootState) => {
  return state.alert ? state.alert.underlyingError || null : null
}
