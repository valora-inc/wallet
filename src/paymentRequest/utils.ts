import { decryptComment, encryptComment } from '@celo/cryptographic-utils'
import { isE164NumberStrict } from '@celo/phone-utils'
import { hexToBuffer } from '@celo/utils/lib/address'
import BigNumber from 'bignumber.js'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import { Recipient } from 'src/recipients/recipient'
import { TransactionDataInput } from 'src/send/SendAmount'
import { TokenBalance } from 'src/tokens/slice'
import Logger from 'src/utils/Logger'
import { Currency } from 'src/utils/currencies'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Convert a payment request into a transaction data object if user has enough balance
export const transactionDataFromPaymentRequest = ({
  paymentRequest,
  stableTokens,
  requester,
}: {
  paymentRequest: PaymentRequest
  stableTokens: TokenBalance[]
  requester: Recipient
}): TransactionDataInput => {
  const cUsdTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Dollar)
  const cEurTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Euro)
  if (!cUsdTokenInfo?.address || !cEurTokenInfo?.address) {
    // Should never happen in production
    throw new Error('No token address found for cUSD or cEUR')
  }
  // If the user has enough cUSD balance, pay with cUSD
  // Else, try with cEUR
  // Else, throw error
  const usdRequested = new BigNumber(paymentRequest.amount)

  if (
    cUsdTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cUsdTokenInfo.balance.multipliedBy(cUsdTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount),
      tokenAmount: new BigNumber(paymentRequest.amount),
      amountIsInLocalCurrency: false,
      tokenAddress: cUsdTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  } else if (
    cEurTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cEurTokenInfo.balance.multipliedBy(cEurTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      tokenAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      amountIsInLocalCurrency: false,
      tokenAddress: cEurTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  }
  throw new Error('Insufficient stablecoin balance')
}

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164NumberStrict(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable') ?? undefined
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
