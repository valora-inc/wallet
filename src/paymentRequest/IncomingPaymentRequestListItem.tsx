import { useFocusEffect } from '@react-navigation/native'
import BigNumber from 'bignumber.js'
import React, { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ActivityIndicator, StyleSheet, View } from 'react-native'
import { useDispatch, useSelector } from 'react-redux'
import { HomeEvents } from 'src/analytics/Events'
import { SendOrigin } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { TokenTransactionType } from 'src/apollo/types'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import RequestMessagingCard from 'src/components/RequestMessagingCard'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { fetchAddressesAndValidate } from 'src/identity/actions'
import { AddressValidationType, SecureSendDetails } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { declinePaymentRequest } from 'src/paymentRequest/actions'
import { Recipient } from 'src/recipients/recipient'
import { RootState } from 'src/redux/reducers'
import { TransactionDataInput } from 'src/send/SendConfirmationLegacy'
import { Currency } from 'src/utils/currencies'
import Logger from 'src/utils/Logger'

interface Props {
  id: string
  requester: Recipient
  amount: string
  comment?: string
}

export default function IncomingPaymentRequestListItem({ id, amount, comment, requester }: Props) {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const [payButtonPressed, setPayButtonPressed] = useState(false)
  const [addressesFetched, setAddressesFetched] = useState(false)

  const e164PhoneNumber = requester.e164PhoneNumber
  const requesterAddress = requester.address

  const secureSendDetails: SecureSendDetails | undefined = useSelector(
    (state: RootState) => state.identity.secureSendPhoneNumberMapping[e164PhoneNumber || '']
  )

  const onPayButtonPressed = () => {
    setPayButtonPressed(true)
    if (e164PhoneNumber) {
      // Need to check latest mapping to prevent user from accepting fraudulent requests
      dispatch(fetchAddressesAndValidate(e164PhoneNumber, requesterAddress))
    } else {
      navigateToNextScreen()
    }

    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.incoming_tx_request,
      selectedAction: NotificationBannerCTATypes.pay,
    })
  }

  const onDeclineButtonPressed = () => {
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.incoming_tx_request,
      selectedAction: NotificationBannerCTATypes.decline,
    })
    dispatch(declinePaymentRequest(id))
    Logger.showMessage(t('requestDeclined'))
  }

  const navigateToNextScreen = () => {
    const transactionData: TransactionDataInput = {
      reason: comment,
      recipient: requester,
      amount: new BigNumber(amount),
      currency: Currency.Dollar,
      type: TokenTransactionType.PayRequest,
      firebasePendingRequestUid: id,
    }

    const addressValidationType =
      secureSendDetails?.addressValidationType || AddressValidationType.NONE

    const origin = SendOrigin.AppRequestFlow
    if (addressValidationType === AddressValidationType.NONE) {
      navigate(Screens.SendConfirmationLegacy, { transactionData, origin })
    } else {
      navigate(Screens.ValidateRecipientIntro, {
        transactionData,
        addressValidationType,
        requesterAddress,
        origin,
      })
    }
  }

  useFocusEffect(
    useCallback(
      () => () => {
        // This is run when focus is lost.
        setPayButtonPressed(false)
        setAddressesFetched(false)
      },
      []
    )
  )

  useEffect(() => {
    // Need this to make sure it's only triggered on click
    if (!payButtonPressed) {
      return
    }

    const isFetchingAddresses = secureSendDetails?.isFetchingAddresses

    if (isFetchingAddresses) {
      setAddressesFetched(true)
    }

    if (addressesFetched && isFetchingAddresses === false) {
      setPayButtonPressed(false)
      if (secureSendDetails?.lastFetchSuccessful) {
        navigateToNextScreen()
      }
    }
  }, [payButtonPressed, secureSendDetails])

  return (
    <View style={styles.container}>
      <RequestMessagingCard
        testID={`IncomingPaymentRequestNotification/${id}`}
        title={t('incomingPaymentRequestNotificationTitle', { name: requester.name })}
        details={comment}
        amount={
          <CurrencyDisplay
            amount={{
              value: amount,
              currencyCode: Currency.Dollar,
            }}
          />
        }
        icon={<ContactCircle recipient={requester} />}
        callToActions={[
          {
            text: payButtonPressed ? (
              <ActivityIndicator testID={'loading/paymentRequest'} />
            ) : (
              t('send')
            ),
            onPress: onPayButtonPressed,
          },
          {
            text: t('decline'),
            onPress: onDeclineButtonPressed,
          },
        ]}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
})
