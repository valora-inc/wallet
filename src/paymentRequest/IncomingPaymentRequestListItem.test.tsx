import { fireEvent, render } from '@testing-library/react-native'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { showError } from 'src/alert/actions'
import { SendOrigin } from 'src/analytics/types'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { AddressValidationType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import IncomingPaymentRequestListItem from 'src/paymentRequest/IncomingPaymentRequestListItem'
import { TransactionDataInput } from 'src/send/SendAmount'
import { createMockStore, getElementText } from 'test/utils'
import {
  mockAccount3,
  mockCeurAddress,
  mockCusdAddress,
  mockE164Number,
  mockPaymentRequests,
  mockPhoneRecipient,
  mockRecipient,
  mockTokenBalances,
} from 'test/values'

const mockGetRecipientFromAddress = jest.fn((address: string) => {
  switch (address) {
    case mockAccount3:
      return mockPhoneRecipient
    default:
      return mockRecipient
  }
})
jest.mock('src/recipients/recipient', () => ({
  ...(jest.requireActual('src/recipients/recipient') as any),
  getRecipientFromAddress: (address: string) => mockGetRecipientFromAddress(address),
}))

const mockPaymentRequest = mockPaymentRequests[1]

const balances = {
  tokenBalances: {
    ...mockTokenBalances,
    [mockCusdAddress]: {
      ...mockTokenBalances[mockCusdAddress],
      balance: '200',
    },
  },
}

const expectedTransactionData: TransactionDataInput = {
  comment: mockPaymentRequest.comment,
  recipient: mockRecipient,
  inputAmount: new BigNumber(mockPaymentRequest.amount),
  tokenAmount: new BigNumber(mockPaymentRequest.amount),
  amountIsInLocalCurrency: false,
  tokenAddress: mockCusdAddress,
  paymentRequestId: mockPaymentRequest.uid,
}

const identityLoading = {
  secureSendPhoneNumberMapping: {
    [mockE164Number]: {
      addressValidationType: AddressValidationType.NONE,
      isFetchingAddresses: true,
      lastFetchSuccessful: true,
    },
  },
}

const identityLoaded = {
  secureSendPhoneNumberMapping: {
    [mockE164Number]: {
      addressValidationType: AddressValidationType.NONE,
      isFetchingAddresses: false,
      lastFetchSuccessful: true,
    },
  },
}

const props = {
  paymentRequest: mockPaymentRequest,
}

describe('IncomingPaymentRequestListItem', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    const store = createMockStore()

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })

  it('shows the phone number if there is no name associated with the address', () => {
    const store = createMockStore()
    const request = {
      ...mockPaymentRequest,
      requesterAddress: mockAccount3,
    }

    const { getByTestId } = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem paymentRequest={request} />
      </Provider>
    )

    expect(getElementText(getByTestId('IncomingPaymentRequestNotification/FAKE_ID_2/Title'))).toBe(
      'incomingPaymentRequestNotificationTitle, {"name":"+15551234567"}'
    )
  })

  it('displays the loading animation while fetching addresses', () => {
    const store = createMockStore()

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )
    fireEvent.press(tree.getByText('send'))
    expect(tree.queryByTestId('loading/paymentRequest')).not.toBeNull()
  })

  it('navigates to send confirmation if there is no validation needed', () => {
    const store = createMockStore({
      identity: identityLoading,
      tokens: balances,
    })

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    fireEvent.press(tree.getByText('send'))

    const updatedStore = createMockStore({
      identity: identityLoaded,
      tokens: balances,
    })

    tree.rerender(
      <Provider store={updatedStore}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(navigate).toHaveBeenCalledWith(Screens.SendConfirmation, {
      origin: SendOrigin.AppRequestFlow,
      transactionData: expectedTransactionData,
      isFromScan: false,
    })
  })

  it('uses cEUR if the cUSD balance is too low', () => {
    const store = createMockStore({
      identity: identityLoading,
      tokens: {
        tokenBalances: {
          ...mockTokenBalances,
          [mockCeurAddress]: {
            ...mockTokenBalances[mockCeurAddress],
            balance: '200',
          },
        },
      },
    })

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    fireEvent.press(tree.getByText('send'))

    const updatedStore = createMockStore({
      identity: identityLoaded,
      tokens: {
        tokenBalances: {
          ...mockTokenBalances,
          [mockCeurAddress]: {
            ...mockTokenBalances[mockCeurAddress],
            balance: '200',
          },
        },
      },
    })

    tree.rerender(
      <Provider store={updatedStore}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(navigate).toHaveBeenCalledWith(Screens.SendConfirmation, {
      origin: SendOrigin.AppRequestFlow,
      transactionData: {
        ...expectedTransactionData,
        tokenAddress: mockCeurAddress,
        tokenAmount: new BigNumber('155.93965517241379310345'),
        inputAmount: new BigNumber('155.93965517241379310345'),
      },
      isFromScan: false,
    })
  })

  it('errors if cUSD and cEUR balance are too low', () => {
    const store = createMockStore({
      identity: identityLoading,
      tokens: {
        tokenBalances: mockTokenBalances,
      },
    })

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    fireEvent.press(tree.getByText('send'))

    const updatedStore = createMockStore({
      identity: identityLoaded,
      tokens: {
        tokenBalances: mockTokenBalances,
      },
    })
    updatedStore.dispatch = jest.fn()

    tree.rerender(
      <Provider store={updatedStore}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(navigate).not.toHaveBeenCalled()
    expect(updatedStore.dispatch).toHaveBeenCalledWith(
      showError(ErrorMessages.INSUFFICIENT_BALANCE_STABLE)
    )
  })

  it('navigates to secure send if there is validation needed', () => {
    const store = createMockStore({
      identity: identityLoading,
      tokens: balances,
    })

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    fireEvent.press(tree.getByText('send'))

    const updatedStore = createMockStore({
      identity: {
        secureSendPhoneNumberMapping: {
          [mockE164Number]: {
            addressValidationType: AddressValidationType.PARTIAL,
            isFetchingAddresses: false,
            lastFetchSuccessful: true,
          },
        },
      },
      tokens: balances,
    })

    tree.rerender(
      <Provider store={updatedStore}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(navigate).toHaveBeenCalledWith(Screens.ValidateRecipientIntro, {
      origin: SendOrigin.AppRequestFlow,
      transactionData: expectedTransactionData,
      addressValidationType: AddressValidationType.PARTIAL,
      requesterAddress: mockRecipient.address,
    })
  })

  it('does not navigate when address fetch is unsuccessful', () => {
    const store = createMockStore({
      identity: identityLoading,
      tokens: balances,
    })

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    fireEvent.press(tree.getByText('send'))

    const updatedStore = createMockStore({
      identity: {
        secureSendPhoneNumberMapping: {
          [mockE164Number]: {
            addressValidationType: AddressValidationType.PARTIAL,
            isFetchingAddresses: false,
            lastFetchSuccessful: false,
          },
        },
      },
      tokens: balances,
    })

    tree.rerender(
      <Provider store={updatedStore}>
        <IncomingPaymentRequestListItem {...props} />
      </Provider>
    )

    expect(navigate).toBeCalledTimes(0)
  })
})
