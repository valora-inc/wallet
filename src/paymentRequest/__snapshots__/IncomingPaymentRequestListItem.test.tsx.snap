// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomingPaymentRequestListItem renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "padding": 16,
        },
        Object {
          "borderRadius": 8,
        },
        undefined,
        Array [
          Object {
            "borderColor": "#E6E6E6",
            "borderWidth": 1,
            "flex": 1,
            "minHeight": 144,
          },
          Object {},
        ],
      ]
    }
    testID="IncomingPaymentRequestNotification/FAKE_ID_2"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 16,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "color": "#666666",
              "fontFamily": "Inter-Regular",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
            }
          }
          testID="IncomingPaymentRequestNotification/FAKE_ID_2/Title"
        >
          incomingPaymentRequestNotificationTitle, {"name":"John Doe"}
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 24,
              "lineHeight": 28,
              "marginBottom": 2,
            }
          }
          testID="IncomingPaymentRequestNotification/FAKE_ID_2/Amount"
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            â‚±
            240.58
          </Text>
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
            }
          }
          testID="IncomingPaymentRequestNotification/FAKE_ID_2/Details"
        >
          My Birthday Present. :) Am I not the best? Celebration. Bam!
        </Text>
      </View>
      <View
        style={Object {}}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "backgroundColor": "hsl(0, 53%, 93%)",
                  "borderRadius": 20,
                  "height": 40,
                  "width": 40,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "hsl(0, 67%, 24%)",
                    "fontSize": 20,
                  },
                ]
              }
            >
              J
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
        }
      }
      testID="IncomingPaymentRequestNotification/FAKE_ID_2/CallToActions"
    >
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="IncomingPaymentRequestNotification/FAKE_ID_2/CallToActions/send/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          send
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="IncomingPaymentRequestNotification/FAKE_ID_2/CallToActions/decline/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          decline
        </Text>
      </View>
    </View>
  </View>
</View>
`;
